File: android/mesh/src/main/java/io/godstone/mesh/identity/WipeDeferredSeams.kt
package io.godstone.mesh.identity

import java.io.File

/**
 * GS-STORE-006: **THE EFFECTFUL SEAMS FOR A RUNTIME THAT DOES NOT YET STAND.**
 *
 * The startup barrier runs BEFORE the runtime object exists -- so at that moment IT OWNS NO PLATFORM RESOURCE: not the
 * transport, not the keystore, not the database handles. THE STARTUP RESUME MUST THEREFORE DEFER **EVERY** EFFECTFUL SEAM,
 * and these are the ones that remain, mirroring the iOS side's four (which were measured there one failure at a time: a
 * deferred filesystem was discovered only after a court's real files were deleted by the live one).
 *
 * EACH ANSWERS **A NAMED, RETRYABLE PENDING** RESULT, WHICH STOPS THE LADDER BEFORE `KEY_ERASED` -- EXACTLY WHERE A PROCESS
 * THAT HAS NOT YET OPENED ITS STORES MUST STOP, AND EXACTLY WHAT THE CARD'S STEP 6 SAYS: "on restart, resume from the
 * durable compatible journal BEFORE opening keys, databases, discovery or a new identity." THE RUNTIME THAT LATER STANDS
 * RESUMES WITH THE LIVE SEAMS, AND BECAUSE THE JOURNAL NOW CARRIES `RUNTIME_DRAINED` AS ITS OWN CHECKPOINT, THE SECOND
 * ATTEMPT CANNOT MISTAKE WORK NOT DONE FOR WORK ALREADY DONE.
 *
 * THE JOURNAL SEAM IS DELIBERATELY **NOT** DEFERRED: writing a checkpoint is DURABILITY, not destruction, and it is the
 * whole point of the startup half -- `WipeJournalDurabilityAdapter` is the journal seam, and it writes.
 */
object WipeDeferredSeams {

    /** The transport of a runtime that does not yet stand: it DRAINS NOTHING and admits nothing. */
    class DeferredTransportRuntimeSeam : TransportRuntimeSeam {
        override fun drainTransport(): RuntimeDrainReceipt = RuntimeDrainReceipt.NotDrained(REASON)
        override fun isQuiesced(): Boolean = false
        override fun fireRadio(msg: String): Boolean = false
        override fun sendVia(msg: String): Boolean = false

        companion object {
            const val REASON = "the runtime does not yet stand: no drain may be performed at the startup barrier"
        }
    }

    /** The vault of a runtime that does not yet stand: EVERY key answereth a named, retryable pending failure. */
    class DeferredKeyVaultSeam : KeyVaultSeam {
        override fun eraseKey(name: String): KeyDeletionResult =
            // NEVER Deleted (a claim about an act nobody performed) and never Absent (a lie about a key nobody asked for).
            KeyDeletionResult.Failed(keyName = name, retryable = true, reason = REASON)

        companion object {
            const val REASON = "the runtime does not yet stand: no key may be erased at the startup barrier"
        }
    }

    /** The identity authority of a runtime that does not yet stand: it PUBLISHES NOTHING, and NAMES its own absence. */
    class DeferredIdentityAuthoritySeam : IdentityAuthoritySeam {
        override fun publishNewIdentity(): String = REASON
        override fun identity(): String? = null

        companion object {
            const val REASON = "the runtime does not yet stand: no identity may be published at the startup barrier"
        }
    }

    /** The artifact filesystem of a runtime that does not yet stand: it DELETES NOTHING -- but it still READS. */
    class DeferredArtifactFileSystemSeam : ArtifactFileSystemSeam {
        /** READING IS NOT AN EFFECT: a reader that could not see the truth would be worse than no reader at all. */
        override fun exists(path: String): Boolean = File(path).exists()

        override fun deleteArtifact(path: String): FileDeletionResult = FileDeletionResult.Failed(path, REASON)

        /** Nothing is readable while no runtime stands: answering `true` here would invite a reader to try. */
        override fun isReadable(path: String): Boolean = false

        companion object {
            const val REASON = "the runtime does not yet stand: no artifact may be deleted at the startup barrier"
        }
    }
}

File: ios/Packages/GodstoneFoundation/Sources/GodstoneMesh/WipeDeferredSeams.swift
import Foundation

/// GS-STORE-006: **THE EFFECTFUL SEAMS FOR A RUNTIME THAT DOES NOT YET STAND.** `MeshRuntime.create` performeth the
/// startup resume BEFORE the runtime object existeth -- so at that moment IT OWNETH NO PLATFORM RESOURCE: not the
/// transport, not the keychain, not the database handles. THE CREATE-TIME RESUME MUST THEREFORE DEFER **EVERY** EFFECTFUL
/// SEAM, not only the transport's, AND THESE TWO ARE THE ONES THAT REMAINED:
///
///   * THE IDENTITY AUTHORITY: the live one calls `MeshIdentity.generateAndStore()` -- and a measured run showed the
///     ladder hanging when a composition that CARRIED A PENDING WIPE was created, the Keychain being precisely the kind of
///     API that BLOCKS rather than fails on the host. THE DEFERRED ONE NEVER TOUCHES IT.
///   * THE KEY VAULT: the live one erases DEKs and identity keys. THE DEFERRED ONE ERASES NOTHING.
///
/// BOTH ANSWER **NAMED, RETRYABLE PENDING** RESULTS, WHICH STOP THE LADDER BEFORE `KEYS_ERASED` -- EXACTLY WHERE A PROCESS
/// THAT HATH NOT YET OPENED ITS STORES MUST STOP, AND EXACTLY WHAT THE CARD'S STEP 6 SAYETH: "on restart, resume from the
/// durable compatible journal BEFORE opening keys, databases, discovery or a new identity." THE RUNTIME THAT LATER STANDS
/// RESUMETH WITH THE LIVE SEAMS, AND BECAUSE THE JOURNAL CARRIETH EACH STAGE, THE SECOND ATTEMPT CANNOT MISTAKE WORK NOT
/// DONE FOR WORK ALREADY DONE.

/// The identity authority of a runtime that does not yet stand. It PUBLISHETH NOTHING.
public final class WipeDeferredIdentityAuthoritySeam: IdentityAuthoritySeam {
    public static let reason = "the runtime does not yet stand: no identity may be published at the startup resume"

    public init() {}

    /// IT NAMETH ITS OWN ABSENCE RATHER THAN INVENTING A NAME: the ladder records `NEW_IDENTITY` only when this is
    /// called, and the caller's journal then carrieth a state that SAYS the identity was not published.
    public func publishNewIdentity() -> String? {
        // IT PUBLISHETH NOTHING, AND NOW IT CAN SAY SO: `nil` stoppeth the ladder AT `ARTIFACTS_DELETED`.
        nil
    }

    /// NO IDENTITY STANDS AT THIS MOMENT -- and `nil` is exactly the seam's own vocabulary for that.
    public func identity() -> String? { nil }
}

/// The key vault of a runtime that does not yet stand. It ERASETH NOTHING.
public final class WipeDeferredKeyVaultSeam: KeyVaultSeam {
    public static let reason = "the runtime does not yet stand: no key may be erased at the startup resume"

    public init() {}

    /// A NAMED, RETRYABLE PENDING FAILURE FOR EVERY KEY -- never `.deleted` (a claim about an act nobody performed) and
    /// never `.absent` (a lie about a key nobody asked for). THE WIPE STAYETH PENDING, which is what the card requireth of
    /// a failed key operation, and `KEYS_ERASED` stayeth unreachable until the runtime that stands owns the vault.
    public func eraseKey(_ name: String) -> KeyDeletionResult {
        return .failed(keyName: name, retryable: true, reason: Self.reason)
    }
}

/// The artifact filesystem of a runtime that does not yet stand. It DELETETH NOTHING -- and IT IS HERE BECAUSE A MEASURED
/// RUN SAID SO: a court which planted a journal at `keyErased` saw this seam (the LIVE one) DELETE ITS REAL PATHS, after
/// which the court could not open its peer store (`stepFailed`, `PeerIdentityStore.swift:292`). THE RULE IS THE SAME RULE
/// A THIRD TIME: at create time the runtime owns no platform resource, so a seam that REMOVES FILES may not run there.
///
/// `exists(_:)` STILL READETH THE REAL FILESYSTEM, AND THAT IS DELIBERATE: READING IS NOT AN EFFECT -- and a reader that
/// could not see the truth would be worse than no reader at all.
public final class WipeDeferredArtifactFileSystemSeam: ArtifactFileSystemSeam {
    public static let reason = "the runtime does not yet stand: no artifact may be deleted at the startup resume"
    private let fileManager: FileManager

    public init(fileManager: FileManager = .default) { self.fileManager = fileManager }

    public func exists(_ path: String) -> Bool { return fileManager.fileExists(atPath: path) }

    /// NEVER `.deleted` and NEVER `.absent`: a NAMED, RETRYABLE PENDING FAILURE, so the ladder STOPPETH before
    /// `ARTIFACTS_DELETED` and the wipe remaineth pending rather than claiming a cleanup it did not perform.
    public func deleteArtifact(_ path: String) -> FileDeletionResult {
        return .failed(path: path, reason: Self.reason)
    }

    /// NOTHING IS READABLE WHILE NO RUNTIME STANDS: the keys that would decrypt anything are not this seam's to judge,
    /// and answering `true` here would invite a reader to try.
    public func isReadable(_ path: String) -> Bool { return false }
}
